Extracted the webpack and build into local webpack config

enejb opened the pull request:
#### Changes proposed in this Pull Request

* Extract the valuable webpack config object from calypso's webpack.config.js and merger it with the build script that is coming from jetpack-blocks package.

#### Testing instructions
* Run 
    ```
    npx lerna run clean
    npm ci
    npx lerna bootstrap --scope='@automattic/jetpack-blocks'
    npx lerna run prepublishOnly --stream --scope='@automattic/jetpack-blocks'
    ```

* copy the files over to your local docker `_inc/blocks`  folder:
    ```
    rsync -a --delete packages/jetpack-blocks/dist/ ../../jetpack/_inc/blocks/
    ```

* does everything still work?

enejb committed:
Extracted the webpack and build into local webpack config

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,185 @@
+				'process.env.NODE_ENV': JSON.stringify( bundleEnv ),
+				global: 'window',
+			} ),
+			new webpack.IgnorePlugin( /^props$/ ), // look into ?
```
This must be stale, isolating removal in https://github.com/Automattic/wp-calypso/pull/31525

sirreal replied:
Confirmed and removed from main webpack config. Let's remove it.

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,185 @@
+ */
+const bundleEnv = config( 'env' );
+const isDevelopment = process.env.DEV_CYCLES === 'true';
+const shouldMinify = false;
```
**Blocker**: We don't want to lose minification!

```suggestion
const shouldMinify = isDevelopment;
```

Seems redundant now…

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,185 @@
+			new webpack.IgnorePlugin( /^\.\/locale$/, /moment$/ ),
+			...SassConfig.plugins( { cssFilename, minify: ! isDevelopment } ),
+			new DuplicatePackageCheckerPlugin(),
+			shouldCheckForCycles &&
```
Let's drop this

lezama replied:
Done

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,185 @@
+// process.env.MINIFY_JS === 'true' ||
+// ( process.env.MINIFY_JS !== 'false' && bundleEnv === 'production' );
+
+const shouldCheckForCycles = process.env.CHECK_CYCLES === 'true';
```
Let's drop this.

```suggestion
```

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,185 @@
+ * Internal variables
+ */
+const bundleEnv = config( 'env' );
+const isDevelopment = process.env.DEV_CYCLES === 'true';
```
`DEV_CYCLES`? This looks off. Maybe:

```suggestion
const isDevelopment = process.env.NODE_ENV !== 'production';
```

sirreal reviewed (dismissed):
Good start, I've given this an initial pass and have suggested some changes.

sirreal committed:
removed from main webpack config

Co-Authored-By: lezama <lezama@gmail.com>

sirreal committed:
use .cache

Co-Authored-By: lezama <lezama@gmail.com>

sirreal committed:
use .cache

Co-Authored-By: lezama <lezama@gmail.com>

sirreal committed:
We don't want to lose minification!

Co-Authored-By: lezama <lezama@gmail.com>

sirreal committed:
NODE_ENV !== 'production'

Co-Authored-By: lezama <lezama@gmail.com>

sirreal committed:
Let's drop this

Co-Authored-By: lezama <lezama@gmail.com>

lezama committed:
don't ignore dist dirs elsewhere

props @sirreal

lezama committed:
don't shouldCheckForCycles for now

lezama commented:
Things seems to build on a second try 👍 

got this error the first time I tried:
```ModuleNotFoundError: Module not found: Error: Can't resolve 'swiper/dist/css/swiper.css' in '~/dev/wp-calypso/packages/jetpack-blocks/src/blocks/slideshow'```

lezama commented on packages/jetpack-blocks/.gitignore:
```diff
@@ -0,0 +1,2 @@
+build/*
+dist/*
```
https://github.com/Automattic/wp-calypso/pull/31503/commits/b427f4b9e841504c7b0f3d21bfe709ce5047266b

lezama commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,185 @@
+/** @format */
```
```suggestion
```

lezama committed:
We don't need @format anymore

lezama commented on packages/jetpack-blocks/bin/build.js:
```diff
@@ -6,107 +6,14 @@ process.env.TARGET_BROWSER = 'true';
-	};
-};
-
 function build() {
```
watch mode?

sirreal replied:
`webpack-cli` probably does it better than we do:

https://webpack.js.org/api/cli/#watch-options

lezama replied:
👍 let's erase it

lezama committed:
Delete build.js

ockham commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,173 @@
+						__dirname,
+						'..',
+						'..',
+						'assets/stylesheets/shared/_utils.scss' // FIX ME
```
@enejb and I were discussing that we need to get rid of this implicit dependency on Calypso here. One option we discussed was to move this (and probably a bunch of dependent files) to a new `calypso-style-utils` package.

ockham replied:
Oh, there's already an issue for this: #31507

Sorry for the noise then :sweat_smile:

lezama committed:
clean whitespaces

lezama committed:
delete build.js, run webpack directly

lezama committed:
Use calypso-build's file-loader

lezama committed:
remove commented code

lezama committed:
Add internal bable.config.js

lezama committed:
Remove external scss dependency

roccotripaldi committed:
add dependency to calypso-color-scheme package

this should fix #31507

ockham commented on packages/jetpack-blocks/babel.config.js:
```diff
@@ -0,0 +1,89 @@
+	] ),
+	overrides: [
+		{
+			test: [ './client/gutenberg/extensions', './packages/jetpack-blocks' ],
```
These relative paths won't work here, but we do (kinda) need these plugins (long story -- it probably works right now because it's using some fallback JSX transform from the React preset instead, but we shouldn't rely on that).

This will be solved by #31464. For now, let's maybe just change this line to 

```suggestion
			test: [ '.src' ],
```

ockham commented on packages/jetpack-blocks/babel.config.js:
```diff
@@ -0,0 +1,89 @@
+		],
+		'@babel/react',
+	],
+	plugins: _.compact( [
```
We can drop the `_.compact()` -- not falsey array items to filter here :slightly_smiling_face:

lezama replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/d544f2b3f5ff756688995033d544fd6d7f549e1b

ockham commented on packages/jetpack-blocks/babel.config.js:
```diff
@@ -0,0 +1,89 @@
+					},
+				],
+			],
+		},
```
I think we can drop the `build_pot` env here since we're not using i18n tools from within here.

lezama replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/384ab7421e1b54d4188ca187ce6a58ea83cb5420

ockham commented on packages/jetpack-blocks/package.json:
```diff
@@ -50,7 +54,7 @@
   "scripts": {
     "clean": "npx rimraf dist",
-    "build": "node bin/build.js",
+    "build": "node $NODE_ARGS ./node_modules/webpack/bin/webpack.js --config ./webpack.config.js",
```
:heart:

sirreal replied:
`webpack-cli` should be a `devDependency`, it's bin will be available, then we should be able to do this:

```suggestion
    "build": "webpack --config ./webpack.config.js",
```

ockham replied:
Even :heart: er

lezama replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/ad8acb90f3239b993ad79804fe3409ff5e2ae98f

ockham commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,155 @@
+				SassConfig.loader( {
+					preserveCssCustomProperties,
+					includePaths: [ path.join( __dirname, 'client' ) ],
+					prelude: `@import '~@automattic/calypso-color-schemes/src/shared/colors';`,
```
:heart:

sirreal replied:
Is there a reason for backtick <code>`</code> quotes here?
```suggestion
					prelude: '@import "~@automattic/calypso-color-schemes/src/shared/colors";',
```

simison replied:
Should we move this import to each file that depends on these colours?

Not necessarily in this PR tho.

ockham reviewed (approved):
This is looking good! Left a few minor suggestions -- I think [this one](https://github.com/Automattic/wp-calypso/pull/31503/files#r266662630) might be the most relevant one.

Since this is contained to the `jetpack-blocks` directory and blocks seem to be working well, I think it's okay to merge this! :shipit: 

(There are a few more `FIXME`s in here that we'll need to sort out before actually moving the code over to the JP repo, but we can sort those out in subsequent PRs.)

ockham committed:
These relative paths won't work here

Co-Authored-By: lezama <lezama@gmail.com>

lezama committed:
Add webpack-cli dependency

props @sirreal

lezama committed:
remove build_pot

props @ockham

lezama committed:
remove .compact

props @ockham

lezama committed:
cleanup

enejb commented:
I tested this locally and I was able to build the dist folder as expected via the command outlined in the test.

enejb commented on packages/jetpack-blocks/package.json:
```diff
@@ -26,7 +26,9 @@
     "@wordpress/token-list": "^1.1.0",
     "@wordpress/url": "^2.3.3",
     "classnames": "^2.2.6",
+    "circular-dependency-plugin": "5.0.2",
```
Can we remove this now?

simison replied:
Removed in https://github.com/Automattic/wp-calypso/pull/31503/commits/da7d36d56f2862d38451315dcd79f2735b44227c

simison commented on packages/jetpack-blocks/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/.cache/
+/dist/
```
Looks like this is already covered by root `.gitignore`:

https://github.com/Automattic/wp-calypso/blob/2b2ca2b01466b33add52c533cb6383cef7080be2/.gitignore#L69-L70

simison replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/b27473f0a65c04955b3d183bb341295c418d919f

simison commented on packages/jetpack-blocks/babel.config.js:
```diff
@@ -0,0 +1,60 @@
+		[
+			'@babel/env',
+			{
+				targets: { browsers: [ 'last 2 versions', 'Safari >= 10', 'iOS >= 10', 'ie >= 11' ] },
```
Watch for https://github.com/Automattic/wp-calypso/pull/31536 in case this needs rebasing.

simison replied:
In fact, I think we can remove this to let Babel rely on default `browserslist` config provided by the consumer (Jetpack).

simison commented on packages/jetpack-blocks/package.json:
```diff
@@ -37,20 +39,23 @@
     "react": "^16.8.3",
-    "webpack": "^4.29.6"
+    "webpack": "^4.29.6",
+    "webpack-cli": "3.2.3"
```
Prolly needs a `^` here.

simison replied:
Actually there's a mixture of strict and `^` here. I think we'll need `/.packages/.npmrc` to override default "strict" setting from the root config.

simison replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/08b8be260d1b12ea564296c97040b2c1383d4d63

simison replied:
>I think we'll need /.packages/.npmrc to override default "strict" setting from the root config.

Nope, not possible. :-( https://docs.npmjs.com/files/npmrc#files

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,155 @@
+			],
+		},
+		resolve: {
+			extensions: [ '.json', '.js', '.jsx' ],
```
For the record, I want the `jsx` extension to go away 🙂 
(no change for this PR)

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,155 @@
+		node: false, // look into ?
+		plugins: _.compact( [
+			new webpack.DefinePlugin( {
+				'process.env.NODE_ENV': JSON.stringify( bundleEnv ),
```
We can likely remove this config dependency:
```suggestion
				'process.env.NODE_ENV': JSON.stringify( process.env.NODE_ENV ),
```

simison replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/7a0909255d2d6dabf6ad4c005fb3366781f4068c

sirreal reviewed (commented):
Coming along nicely, a few more notes 🙂

simison committed:
Remove reduntant dist from gitignore

simison committed:
Use correct license SPDX

"GPL-2.0-or-later" is the new one, "GPL-2.0+" is deprecated.

https://spdx.org/licenses/

simison committed:
Relax semver

sirreal committed:
Update Sass config prelude not use backticks

Co-Authored-By: simison <mikael@ihminen.org>

simison committed:
Remove `bundleEnv`

simison commented:
Desktop tests are failing with:
```
Cloning into '.'...
Warning: Permanently added the RSA host key for IP address '192.30.253.112' to the list of known hosts.

remote: Enumerating objects: 3596, done.        
remote: Total 3596 (delta 0), reused 0 (delta 0), pack-reused 3596        
Receiving objects: 100% (3596/3596), 16.07 MiB | 0 bytes/s, done.
Resolving deltas: 100% (2070/2070), done.
fatal: Couldn't find remote ref tests/add/webpack-jetpack-blocks
Exited with code 128
```
What's going on? 🤔

simison committed:
Remove `circular-dependency-plugin`

sirreal commented:
> Desktop tests are failing with:

Intermittent failure. Green again 👍

sirreal commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,153 @@
+		context: __dirname,
+		entry: {
+			editor: editorScript,
+			...( editorBetaScript && { 'editor-beta': editorBetaScript } ),
```
We expect to always have an `editor-beta` now:
```suggestion
			'editor-beta': editorBetaScript,
```

simison replied:
Yep, we'll always want to build beta files, even when beta list is empty.

sirreal committed:
Always expect beta build

Co-Authored-By: simison <mikael@ihminen.org>

simison commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,153 @@
+			new webpack.IgnorePlugin( /^\.\/locale$/, /moment$/ ),
+			...SassConfig.plugins( { cssFilename, minify: ! isDevelopment } ),
+			new DuplicatePackageCheckerPlugin(),
+			fs.existsSync( presetPath ) &&
```
This file should always exist now.

simison replied:
https://github.com/Automattic/wp-calypso/pull/31503/commits/d6bfe0656f68ee39266797a5fa28ec816f09cd00

simison commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,153 @@
+			extensions: [ '.json', '.js', '.jsx' ],
+			modules: [ 'node_modules' ],
+		},
+		node: false, // look into ?
```
```js
// look into ?
```

Flagging for visibility. :-)

sirreal replied:
I think we're good:

https://webpack.js.org/configuration/node/#node

https://github.com/Automattic/wp-calypso/blob/05fec274b13cb1a3603a591cd565bdebb4660190/webpack.config.js#L215

```suggestion
		node: false,
```

simison committed:
Change versions back to strict

...since Jetpack blocks are an app, and not a library and using strict versions will help deduping easier if these versions match Calypso versions.

simison committed:
Always expect preset file to exist

simison committed:
Minor cleanup

simison reviewed (approved):
Looks good now and builds well, tested in the editor with `JETPACK_BETA_BLOCKS` set `true` & `false`.

simison commented:
This could do with final +1 since there were a bunch of commits by me and Jon after Bernie's ✅ review.

sirreal committed:
Remove todo

Co-Authored-By: simison <mikael@ihminen.org>

simison commented:
I see wrong colours in Jetpack sidebar toggle logo:

<img width="333" alt="Screenshot 2019-03-19 at 12 20 37" src="https://user-images.githubusercontent.com/87168/54598870-258ff880-4a42-11e9-9964-a5e590da24d9.png">
<img width="1156" alt="Screenshot 2019-03-19 at 12 20 26" src="https://user-images.githubusercontent.com/87168/54598871-26288f00-4a42-11e9-94c9-dc1139185bc2.png">

Colours are fine in `master`:

<img width="268" alt="image" src="https://user-images.githubusercontent.com/87168/54600793-3b072180-4a46-11e9-8177-867d0529efea.png">

It's from here:

https://github.com/Automattic/wp-calypso/blob/bd2c8ac6d0c223f5fdc5aaa4071db86c3fa98916/packages/jetpack-blocks/src/shared/jetpack-plugin-sidebar.scss#L31-L39

In `master` we get this CSS:

```css
.edit-post-pinned-plugins .components-icon-button:not(.is-toggled) .jetpack-logo__icon-triangle {
    fill: #fff !important;
}
```

In this branch following:
```css
.edit-post-pinned-plugins .components-icon-button:hover .jetpack-logo__icon-triangle, .edit-post-pinned-plugins .components-icon-button:not(.is-toggled) .jetpack-logo__icon-triangle {
    fill: #fff !important;
    fill: var(--color-white) !important;
}
```

We rely quite a bit on these `--color-*` variables in our blocks so I'd expect other places to be broken as well.

simison committed:
Rebase with prettier tabs

simison commented:
Rebased.

simison committed:
Create package-lock.json

Since jetpack-blocks isn’t ref’d from the root, it doesn’t participate in the root npm-shrinkwrap

simison commented on packages/jetpack-blocks/package.json:
```diff
@@ -8,49 +8,53 @@
-		"webpack": "^4.29.6"
+		"@automattic/calypso-build": "file:../calypso-build",
+		"@automattic/calypso-color-schemes": "file:../calypso-color-schemes",
+		"chalk": "2.4.2",
```
I can't see where `chalk` is used?

sirreal replied:
It was in `bin/build.js`
```suggestion
```

simison replied:
Removed. Will need to re package-lock (unless we remove the whole file https://github.com/Automattic/wp-calypso/pull/31503#issuecomment-474333824).

simison commented:
I added `package-lock.json` as per comment:

> Mikael i think you want package locks there. Since jetpack-blocks isn’t ref’d from the root, it doesn’t participate in the root npm-shrinkwrap

(via p1552993298168400-slack-team-calypso)

and now it seems like build is failing:

https://circleci.com/gh/Automattic/wp-calypso/221156?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Locally, running `npx lerna bootstrap --scope='@automattic/jetpack-blocks'` did pass tho.

sirreal committed:
Remove `chalk` dep

Co-Authored-By: simison <mikael@ihminen.org>

simison reviewed (dismissed):
So current PR status:

- [x] ~Colours need to be fixed. see https://github.com/Automattic/wp-calypso/pull/31503#issuecomment-474314700~ fixed in https://github.com/Automattic/wp-calypso/pull/31503/commits/e034860e64f7eb1b980b4fa09f1e24cfeaf07a46
- [x] ~package-lock removed? see https://github.com/Automattic/wp-calypso/pull/31503#issuecomment-474333824~ fixed in https://github.com/Automattic/wp-calypso/pull/31503/commits/432ed69b6234e9a410a01d21415eb89cc53bf5da
- [x] ~figure out if `cacheIdentifier` should be removed; https://github.com/Automattic/wp-calypso/pull/31503#discussion_r266860353~ removed in https://github.com/Automattic/wp-calypso/pull/31503/commits/13ed79a49fdaa6954864669dc8ed99ebc34228d7

simison commented on packages/jetpack-blocks/webpack.config.js:
```diff
@@ -0,0 +1,151 @@
+/**
+ * Internal dependencies
+ */
+const cacheIdentifier = require( '../../server/bundler/babel/babel-loader-cache-identifier' ); // FIX ME
```
This can probably be removed but I suppose that can be a separate PR as well.

Convo p1552997442520300-slack-jetpack-gutenberg

sirreal replied:
I'm digging into this right now. Will report back…

sirreal replied:
Let's drop it.

[Babel-loader default](https://github.com/babel/babel-loader/blob/5da4fa0c673474ab14bc829f2cbe03e60e437858/src/index.js#L178-L182):

```js
      const cacheIdentifier = JSON.stringify({
        options,
        "@babel/core": transform.version,
        "@babel/loader": pkg.version,
      });
```

Our implementation:
https://github.com/Automattic/wp-calypso/blob/7a09dd2cf8219188bdc9c3e668cc8c34108169bc/server/bundler/babel/babel-loader-cache-identifier/index.js#L28-L36

It's essentially the same with the addition of our custom babel async transform which we aren't (I hope) bringing with us.

sirreal committed:
Regenerate package-lock with lerna

```
npx lerna bootstrap --scope='@automattic/jetpack-blocks' -- --save
```

sirreal committed:
Improve CI build

sirreal committed:
Drop cacheIdentifier

simison committed:
Do not preserve custom CSS properties

sirreal reviewed (approved):
If it builds and works, let's ship this. It looks good 👍

simison commented:
Something still failing with the build. :-/

sirreal committed:
Remove jetpack-blocks package-lock

sirreal committed:
Revert "Improve CI build"

This reverts commit d337992a82a45ed770cba7fb366b9b782d3bf3a6.

sirreal reviewed (approved):
Tests are all green, let's ship
